### PR TITLE
feat(list): Add single selection

### PR DESCRIPTION
**Pull request by williamernest**

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -6176,6 +6176,7 @@
           "version": "0.0.9",
           "bundled": true,
           "dev": true,
+          "optional": true,
```
Comment (patrickrodee) [inline package-lock.json]: Was this file supposed to change?

File: packages/mdc-list/README.md [hunk 2]
```diff
@@ -212,7 +254,11 @@ Method Signature | Description
 `getListItemCount() => Number` | Returns the total number of list items (elements with `mdc-list-item` class) that are direct children of the `root_` element.
 `getFocusedElementIndex() => Number` | Returns the `index` value of the currently focused element.
 `getListItemIndex(ele: Element) => Number` | Returns the `index` value of the provided `ele` element.
+`setAttributeForElementIndex(ndx: Number, attr: String, value: String) => void` | Sets the `attr` attribute to the value of `value` for the list item at `ndx`.
```
Comment (patrickrodee) [inline packages/mdc-list/README.md]: I'm not in love with the variable name `ndx` but it's already in here so I'm not gonna request a change.

File: packages/mdc-list/foundation.js [hunk 3]
```diff
@@ -63,6 +79,43 @@ class MDCListFoundation extends MDCFoundation {
...
+  setSingleSelection(value) {
+    this.isSingleSelectionList_ = value;
+  }
+
+  /** @param {number} ndx */
+  setSelectedIndex(ndx) {
+    if (ndx === this.selectedIndex_) {
+      this.adapter_.setAttributeForElementIndex(this.selectedIndex_, strings.ARIA_SELECTED, false);
+      this.adapter_.removeClassForElementIndex(this.selectedIndex_, cssClasses.LIST_SELECTED_CLASS);
+      if (this.selectedIndex_ > 0) {
```
Comment (patrickrodee) [inline packages/mdc-list/foundation.js]: Why does this only set negative tabindex for items 1...n instead of 0...n?

File: packages/mdc-list/foundation.js [hunk 4]
```diff
@@ -63,6 +79,43 @@ class MDCListFoundation extends MDCFoundation {
...
+    }
+
+    if (this.selectedIndex_ >= 0) {
+      this.adapter_.setAttributeForElementIndex(this.selectedIndex_, strings.ARIA_SELECTED, false);
+      this.adapter_.removeClassForElementIndex(this.selectedIndex_, cssClasses.LIST_SELECTED_CLASS);
+      this.adapter_.setAttributeForElementIndex(this.selectedIndex_, 'tabindex', -1);
+    }
+
+    if (ndx >= 0) {
+      if (this.adapter_.getListItemCount() > ndx) {
```
Comment (patrickrodee) [inline packages/mdc-list/foundation.js]: Collapse these into one if statement

File: packages/mdc-list/foundation.js [hunk 5]
```diff
@@ -120,7 +176,37 @@ class MDCListFoundation extends MDCFoundation {
     } else if (isEnd) {
       this.preventDefaultEvent_(evt);
       this.focusLastElement();
+    } else if (this.isSingleSelectionList_ && (isEnter || isSpace)) {
+      this.preventDefaultEvent_(evt);
+      // Check if the space key was pressed on the list item or a child element.
+      if (evt.target.classList.contains(cssClasses.LIST_ITEM_CLASS)) {
```
Comment (patrickrodee) [inline packages/mdc-list/foundation.js]: This should be an adapter method (`hasClass` or something similar)

File: packages/mdc-list/foundation.js [hunk 6]
```diff
@@ -120,7 +176,37 @@ class MDCListFoundation extends MDCFoundation {
...
+
+    if (currentIndex === -1) {
+      const listItem = this.getListItem_(evt.target);
+      if (this.adapter_.isElementFocusable(evt.target)) {
+        return;
+      }
+
+      currentIndex = this.adapter_.getListItemIndex(listItem);
+
+      if (currentIndex < 0) {
```
Comment (patrickrodee) [inline packages/mdc-list/foundation.js]: This if statement can be removed since it always evaluates to true.

File: packages/mdc-list/index.js [hunk 7]
```diff
@@ -89,13 +94,47 @@ export class MDCList extends MDCComponent {
...
+  /** @param {boolean} value */
+  set singleSelection(value) {
+    if (value) {
+      this.root_.addEventListener('click', this.handleClick_);
+    } else {
+      this.root_.removeEventListener('click', this.handleClick_);
+    }
+
+    this.foundation_.setSingleSelection(value);
+    const selectedElement = this.root_.querySelector('.mdc-list-item[aria-selected="true"]');
```
Comment (patrickrodee) [inline packages/mdc-list/index.js]: Could you provide some explanation about why this uses the aria-selected attribute instead of the `mdc-list-item--selected` class?

File: packages/mdc-list/index.js [hunk 8]
```diff
@@ -89,13 +94,47 @@ export class MDCList extends MDCComponent {
...
+    [].slice.call(this.root_.querySelectorAll('.mdc-list-item:not([aria-selected="true"])'))
```
Comment (patrickrodee) [inline packages/mdc-list/index.js]: Could you provide some explanation about why this uses the aria-selected attribute instead of the `mdc-list-item--selected` class?

File: packages/mdc-list/index.js [hunk 9]
```diff
@@ -89,13 +94,47 @@ export class MDCList extends MDCComponent {
...
+      addClassForElementIndex: (ndx, className) => this.listElements_[ndx].classList.add(className),
+      removeClassForElementIndex: (ndx, className) => this.listElements_[ndx].classList.remove(className),
       focusItemAtIndex: (ndx) => this.listElements_[ndx].focus(),
+      isElementFocusable: (ele) => {
+        if (!ele) return false;
+        let matches = Element.prototype.matches;
+        if (!matches) { // IE uses a different name for the same functionality
+          matches = Element.prototype.msMatchesSelector;
+        }
+        return matches.call(ele, strings.FOCUSABLE_CHILD_ELEMENTS);
```
Comment (patrickrodee) [inline packages/mdc-list/index.js]: Clever!!!

File: packages/mdc-list/README.md [hunk 10]
```diff
@@ -135,6 +135,44 @@ OR
...
+```js
+var listEle = document.getElementById('my-list');
+var list = new mdc.list.MDCList(listEle);
+list.singleSelection = true;
+```
+
+#### Pre-selected list item
+
+When rendering the list with a pre-selected list item, the list item that needs to be selected should contain
+the `mdc-list-item--selected` class before creating the list.
```
Comment (patrickrodee) [inline packages/mdc-list/README.md]: Should the list item also have the `aria-selected="true"` attribute?

Comment (williamernest) [reply packages/mdc-list/README.md]: No, it gets added automatically when the list component is created. We add the `--selected` class because we use that as our selector.

File: packages/mdc-list/index.js [hunk 11]
```diff
@@ -89,13 +94,48 @@ export class MDCList extends MDCComponent {
     this.foundation_.setWrapFocus(value);
   }
 
+  /** @param {boolean} value */
+  set singleSelection(value) {
```
Comment (patrickrodee) [inline packages/mdc-list/index.js]: Use a more descriptive variable name than `value` like `isSelected` or something similar.

File: packages/mdc-list/README.md (hunk 10 above)
Comment (patrickrodee) [reply packages/mdc-list/README.md]: Gotcha. Can you show me the line number where that occurs?

Comment (williamernest) [reply packages/mdc-list/README.md]: https://github.com/material-components/material-components-web/blob/1d5f6106e02a3755ceb9eccaa2a0bf39444d08e8/packages/mdc-list/index.js#L106

Comment (patrickrodee) [reply packages/mdc-list/README.md]: I don't see where `aria-selected="true"` gets added though. Could you point me to that?

Comment (williamernest) [reply packages/mdc-list/README.md]: That's in the foundations `setSelectedIndex`.

Comment (williamernest) [reply packages/mdc-list/README.md]: If the `selectedElement` is found it sets the `selectedIndex` to the `selectedElement` index. The `selectedIndex` proxies to the foundation `setSelectedIndex`, which sets the attributes on the list element. 

https://github.com/material-components/material-components-web/blob/cbbd3948ee793abddb5bdbdaa8208385f5e7addd/packages/mdc-list/index.js#L111

https://github.com/material-components/material-components-web/blob/cbbd3948ee793abddb5bdbdaa8208385f5e7addd/packages/mdc-list/index.js#L117

https://github.com/material-components/material-components-web/blob/cbbd3948ee793abddb5bdbdaa8208385f5e7addd/packages/mdc-list/foundation.js#L115

Comment (patrickrodee) [reply packages/mdc-list/README.md]: Awesome, thanks for the patience and guiding me through. Since that means the `aria-selected` attribute is only set by the MDC component index.js implementation during initialization, let's add a note to include `aria-selected="true"` for the pre-selected list item. That will help users who wrap our components and our users in general.

Comment (williamernest) [reply packages/mdc-list/README.md]: ? `aria-selected` is set by the foundation

Comment (williamernest) [reply packages/mdc-list/README.md]: If you wrap the component and want to use the single selection list, you have to call `setSelectedIndex` on the foundation so that it can initialize the initial selected value. `setSelectedIndex` also sets `aria-selected` on the list element that is selected.

File: packages/mdc-list/adapter.js [hunk 12]
```diff
@@ -41,11 +41,48 @@ class MDCListAdapter {
   /** @param {Element} node */
   getListItemIndex(node) {}
 
+  /**
+   * @param {Number} index
```
Comment (patrickrodee) [inline packages/mdc-list/adapter.js]: These basic types (number, string, boolean, etc) should be lowercase.

File: packages/mdc-list/constants.js [hunk 13]
```diff
@@ -18,14 +18,16 @@
...
 };
 
 /** @enum {string} */
 const strings = {
   ARIA_ORIENTATION: 'aria-orientation',
   ARIA_ORIENTATION_VERTICAL: 'vertical',
+  ARIA_SELECTED: 'aria-selected',
   FOCUSABLE_CHILD_ELEMENTS: 'button:not(:disabled), a',
-  ITEMS_SELECTOR: '.mdc-list-item',
+  ITEMS_SELECTOR: '.mdc-list-item:not(.mdc-list-item--disabled)',
```
Comment (patrickrodee) [inline packages/mdc-list/constants.js]: Why did you add the `:not(.mdc-list-item--disabled)` selector?

File: packages/mdc-list/index.js [hunk 14]
```diff
@@ -80,22 +85,55 @@ export class MDCList extends MDCComponent {
...
 
   /** @param {boolean} value */
   set wrapFocus(value) {
     this.foundation_.setWrapFocus(value);
   }
 
+  /** @param {boolean} isSingleSelectionList */
+  set singleSelection(isSingleSelectionList) {
+    if (isSingleSelectionList) {
+      this.root_.addEventListener('click', this.handleClick_);
```
Comment (patrickrodee) [inline packages/mdc-list/index.js]: This should be a foundation method.

File: packages/mdc-list/constants.js (hunk 13 above)
Comment (williamernest) [reply packages/mdc-list/constants.js]: Disabled items should not be receive focus, so the items selector should ignore them.

File: packages/mdc-list/index.js (hunk 14 above)
Comment (williamernest) [reply packages/mdc-list/index.js]: ? Aren't we moving toward adding event listeners in the index.js ? See `initialSyncWithDOM` above.

Comment (patrickrodee) [reply packages/mdc-list/index.js]: True, we're exploring it. I guess we can try it here and see how it goes.

Will `handleClick` cause problems if it's triggered by a non-single selection list item? If so, there should be a check inside the `handleClick` method that ensures it won't cause problems if it's called.

Comment (williamernest) [reply packages/mdc-list/index.js]: `handleClick` has logic to determine if you did not click on a list item (it looks for an ancestor `mdc-list-item`) and if doesn't find one it returns early.

**Comment by trimox** [comment]
Hey @williamernest I opened PR #3029 to export `MDCListFoundation`. Would you prefer to add those exports here? If so, I'll close #3029.

File: packages/mdc-list/README.md [hunk 15]
```diff
@@ -200,30 +239,76 @@ Key | Action
...
+`setSelectedIndex()` should be called with the initially selected element immediately after the foundation is 
+instantiated.
+
+```html
+<ul id="my-list" class="mdc-list" aria-orientation="vertical">
+  <li class="mdc-list-item" tabindex="-1">Single-line item<button tabindex="-1"></button></li>
+  <li class="mdc-list-item mdc-list-item--selected" aria-selected="true" tabindex="0">Single-line item</li>
+  <li class="mdc-list-item" tabindex="-1">Single-line item</li>
+</ul>
+```
```
Comment (moog16) [inline packages/mdc-list/README.md]: I think adding the JS here showing to call `setSelectedIndex` will complete the documentation....if I'm reading that correctly.

File: packages/mdc-list/constants.js [hunk 16]
```diff
@@ -18,14 +18,16 @@
 /** @enum {string} */
 const cssClasses = {
   LIST_ITEM_CLASS: 'mdc-list-item',
+  LIST_SELECTED_CLASS: 'mdc-list-item--selected',
```
Comment (moog16) [inline packages/mdc-list/constants.js]: LIST_ITEM_SELECTED_CLASS is more consistent with the above class

File: packages/mdc-list/constants.js (hunk 13 above)
Comment (moog16) [reply packages/mdc-list/constants.js]: Changing the key to `ENABLED_ITEMS_SELECTOR` is more semantically correct

File: packages/mdc-list/foundation.js [hunk 17]
```diff
@@ -120,9 +190,26 @@ class MDCListFoundation extends MDCFoundation {
     } else if (isEnd) {
       this.preventDefaultEvent_(evt);
       this.focusLastElement();
+    } else if (this.isSingleSelectionList_ && (isEnter || isSpace)) {
```
Comment (moog16) [inline packages/mdc-list/foundation.js]: I don't see a demo for this, and this seems like an important thing to demo.

File: packages/mdc-list/index.js [hunk 18]
```diff
@@ -80,27 +85,62 @@ export class MDCList extends MDCComponent {
...
-    return new MDCListFoundation(/** @type {!MDCListAdapter} */{
+    return new MDCListFoundation(/** @type {!MDCListAdapter} */ (Object.assign({
       getListItemCount: () => this.listElements_.length,
       getFocusedElementIndex: () => this.listElements_.indexOf(document.activeElement),
       getListItemIndex: (node) => this.listElements_.indexOf(node),
-      focusItemAtIndex: (ndx) => this.listElements_[ndx].focus(),
+      setAttributeForElementIndex: (index, attr, value) => this.listElements_[index].setAttribute(attr, value),
+      removeAttributeForElementIndex: (index, attr) => this.listElements_[index].removeAttribute(attr),
+      addClassForElementIndex: (index, className) => this.listElements_[index].classList.add(className),
+      removeClassForElementIndex: (index, className) => this.listElements_[index].classList.remove(className),
```
Comment (moog16) [inline packages/mdc-list/index.js]: I checked the foundation and it looks like you have null check code there (eg. this.selectedIndex >= 0), but is there a possibility that `this.listElements_[index]` could be undefined? Just wanna check

File: test/unit/mdc-list/mdc-list.test.js [hunk 19]
```diff
@@ -72,6 +78,43 @@ test('#adapter.getListItemIndex returns the index of the element specified', ()
...
+  assert.isTrue(selectedNode.classList.contains('foo'));
+  document.body.removeChild(root);
+});
+
+test('#adapter.removeClassForElementIndex removes the class from the list element at index specified', () => {
+  const {root, component} = setupTest();
+  document.body.appendChild(root);
+  const selectedNode = root.querySelectorAll('.mdc-list-item')[1];
+  component.getDefaultFoundation().adapter_.removeClassForElementIndex(1, 'foo');
+  assert.isFalse(selectedNode.classList.contains('foo'));
```
Comment (moog16) [inline test/unit/mdc-list/mdc-list.test.js]: when does the element get the foo class? I ran this and it looks like if you remove line 113 it still passes

File: test/unit/mdc-list/foundation.test.js [hunk 20]
```diff
@@ -383,7 +392,197 @@ test('#handleKeydown does not find ancestor with mdc-list-item so returns early'
...
+  const event = {key: 'Space', target, preventDefault};
+
+  td.when(mockAdapter.getFocusedElementIndex()).thenReturn(0);
+  td.when(mockAdapter.getListItemCount()).thenReturn(3);
+  td.when(mockAdapter.isListItem(target)).thenReturn(true);
+  foundation.setSingleSelection(true);
+  foundation.handleKeydown(event);
+
+  td.verify(preventDefault(), {times: 1});
+  td.verify(mockAdapter.setAttributeForElementIndex(0, strings.ARIA_SELECTED, true));
```
Comment (moog16) [inline test/unit/mdc-list/foundation.test.js]: do you want a {times: 1} here?

File: test/unit/mdc-list/foundation.test.js [hunk 21]
```diff
@@ -383,7 +392,197 @@ test('#handleKeydown does not find ancestor with mdc-list-item so returns early'
...
+
+  td.when(mockAdapter.getFocusedElementIndex()).thenReturn(0);
+  td.when(mockAdapter.getListItemCount()).thenReturn(3);
+  td.when(mockAdapter.isListItem(target)).thenReturn(true);
+  foundation.setSingleSelection(true);
+  foundation.handleKeydown(event);
+  foundation.handleKeydown(event);
+
+  td.verify(preventDefault(), {times: 2});
+  td.verify(mockAdapter.removeAttributeForElementIndex(0, strings.ARIA_SELECTED));
```
Comment (moog16) [inline test/unit/mdc-list/foundation.test.js]: here too {times: 1}. Few places below as well.

File: test/unit/mdc-list/foundation.test.js [hunk 22]
```diff
@@ -383,7 +392,197 @@ test('#handleKeydown does not find ancestor with mdc-list-item so returns early'
...
+  td.verify(preventDefault(), {times: 2});
+  td.verify(mockAdapter.setAttributeForElementIndex(0, 'tabindex', 0));
+  td.verify(mockAdapter.setAttributeForElementIndex(1, 'tabindex', -1));
+});
+
+test('#handleKeydown space key is triggered and focused is moved to a different element', () => {
+  const {foundation, mockAdapter} = setupTest();
+  const preventDefault = td.func('preventDefault');
+  const target = {classList: ['mdc-list-item']};
+  const event = {key: 'Space', target, preventDefault};
```
Comment (moog16) [inline test/unit/mdc-list/foundation.test.js]: You don't have any tests events with keyCodes for space and enter. Will that increase coverage?

File: test/unit/mdc-list/mdc-list.test.js [hunk 23]
```diff
@@ -115,6 +190,33 @@ test('wrapFocus calls setWrapFocus on foundation', () => {
...
+  td.verify(mockFoundation.handleClick(td.matchers.anything()), {times: 0});
+});
+
+test('singleSelection calls foundation setSingleSelection with the provided value', () => {
+  const {component, mockFoundation} = setupTest();
+  component.singleSelection = true;
+  td.verify(mockFoundation.setSingleSelection(true), {times: 1});
+});
+
+test('selectedIndex calls setSelectedIndex on foundation', () => {
```
Comment (moog16) [inline test/unit/mdc-list/mdc-list.test.js]: setSelectedIndex is a pretty complex function, but only has this one test. Might want to add more

File: test/unit/mdc-list/foundation.test.js (hunk 20 above)
Comment (williamernest) [reply test/unit/mdc-list/foundation.test.js]: `{times: 1}` is the default behavior.

File: packages/mdc-list/README.md [hunk 24]
```diff
@@ -200,30 +239,76 @@ Key | Action
...
+
+The `MDCListFoundation` expects the HTML to be setup a certain way before being used. This setup is a part of the `layout()` and `singleSelection()` functions within the `index.js`.
+
+#### Setup in `layout()`
+
+The default component requires that every list item receives a `tabindex` value so that it can receive focus
+(`li` elements cannot receive focus at all without a `tabindex` value). Any element not already containing a
+`tabindex` attribute will receive `tabindex=-1`. The first list item should have `tabindex="0"` so that it the
+user can find the first element using the `tab` key, but subsequent `tab` keys strokes will cause focus to
+skip over the entire list. If the list items contain sub-elements that are focusable (`button` or `a` elements),
```
[the remark this answers is not in the document]
Comment (moog16) [reply packages/mdc-list/README.md]: oh sorry - i meant to remove this comment.

File: test/unit/mdc-list/foundation.test.js (hunk 20 above)
Comment (kfranqueiro) [reply test/unit/mdc-list/foundation.test.js]: I'm pretty sure that is not correct. The default behavior is any non-zero number of invocations is valid.

File: test/unit/mdc-list/foundation.test.js (hunk 22 above)
Comment (williamernest) [reply test/unit/mdc-list/foundation.test.js]: It doesn't doesn't increase coverage since that statement is already covered by the `key` parameter.

File: test/unit/mdc-list/mdc-list.test.js (hunk 23 above)
Comment (williamernest) [reply test/unit/mdc-list/mdc-list.test.js]: `setSelectedIndex` has nearly 100% coverage through the `click`/`key` handler tests.

File: packages/mdc-list/README.md (hunk 15 above)
Comment (williamernest) [reply packages/mdc-list/README.md]: The documentation shows the JS in above examples. This example is discussing how the html should be rendered when creating your own component.

File: test/unit/mdc-list/foundation.test.js [hunk 25]
```diff
@@ -383,7 +419,230 @@ test('#handleKeydown does not find ancestor with mdc-list-item so returns early'
...
+
+  td.when(mockAdapter.getFocusedElementIndex()).thenReturn(0);
+  td.when(mockAdapter.getListItemCount()).thenReturn(3);
+  foundation.setSingleSelection(true);
+  foundation.handleKeydown(event);
+
+  td.verify(preventDefault(), {times: 1});
+});
+
+test('#handleKeydown enter key causes preventDefault to be called on the event', () => {
```
Comment (moog16) [inline test/unit/mdc-list/foundation.test.js]: I think you have the titles of this and the next tests reversed with regards to `if singleSelection=true`

File: test/unit/mdc-list/foundation.test.js [hunk 26]
```diff
@@ -383,7 +419,230 @@ test('#handleKeydown does not find ancestor with mdc-list-item so returns early'
...
+  const event = {key: 'Space', target, preventDefault};
+
+  td.when(mockAdapter.getFocusedElementIndex()).thenReturn(0);
+  td.when(mockAdapter.getListItemCount()).thenReturn(3);
+  foundation.handleKeydown(event);
+
+  td.verify(preventDefault(), {times: 0});
+});
+
+test('#handleKeydown enter key does not causes preventDefault to be called if singleSelection=true', () => {
```
Comment (moog16) [inline test/unit/mdc-list/foundation.test.js]: same here - titles reversed
